### PR TITLE
Close #18774: Migrate mutli-selection to store

### LINKED ISSUE #18774: Migrate mutli-selection to store
We tried the `recyclerview-selection` library and while it worked, it was difficult to use.

- We had to keep a track of the selections in order to know when entering/exiting selection mode.
- The selection was tied tightly to the RecyclerView - this is obvious in hind sight (🤦 ) but when setting up the selection tracker and retrieving the selected items to be used in parent `ViewGroup`, we were re-creating a selection storage concept all over again.

This is a problem that's already solved in the "library" by using the `SelectionInteractor` and `SelectionHolder` along with `lib-state`.

In #17821, we added the storer, so let's put our multi-select items in that instead.

### PR DESCRIPTION
**Pull request by jonalmeida**
**(Meta)** The PR has multiple pieces here, so it would be easier to review it per-commit:
 1. Depends on https://github.com/mozilla-mobile/fenix/pull/18773.
 2. Moves the `SelectionInteractor` to a new package to be shareable.
 3. Replaces the `recyclerview-selection` library with the interactor.
 4. Removes some constructor args from viewholders to fix detekt warnings.

---

This is an implementation that's similar to what we have in other UI lists (library) and works much better with our current architecture.

Visually, it functions just like the other patch, but now we have the ability to enter/exit selection mode, trigger sync, and observe the selected page all from one store. 🙃 

Fixes #18774 

https://user-images.githubusercontent.com/1370580/113471523-ec5d3200-942a-11eb-89cc-ef82cdb86c6e.mp4

### Pull Request checklist

- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

File: app/src/main/java/org/mozilla/fenix/tabstray/TabsTrayStore.kt [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+/**
+ * Value type that represents the state of the tabs tray.
+ *
+ * @property selectedPage The current page in the tray can be on.
+ * @property mode Whether the browser tab list is in multi-select mode or not with the set of
+ * currently selected tabs.
+ * @property syncing Whether the Synced Tabs feature should fetch the latest tabs from paired
+ * devices.
+ */
+data class TabsTrayState(
```
Comment (rocketsroger) [inline app/src/main/java/org/mozilla/fenix/tabstray/TabsTrayStore.kt:21]: I am assuming the `TabsTrayStore` change will land first?

**Comment by jonalmeida** [comment]
Rebased now that the dependant PR is merged.

**Comment by jonalmeida** [comment]
`test-debug` failed from unrelated intermittent test failures: 

- `DefaultToolbarMenuTest. WHEN the bottom toolbar is set THEN the first item in the list is not the navigation`
- `DefaultToolbarMenuTest. WHEN the bottom toolbar is set THEN the nav menu should be the last item`
- `DefaultToolbarMenuTest. WHEN the top toolbar is set THEN settings should be the last item`
- `DefaultToolbarMenuTest. WHEN the top toolbar is set THEN the first item in the list is the navigation`

They all had the same stack trace:

<details>

```
java.lang.ClassCastException: android.content.Context$Subclass0 cannot be cast to org.mozilla.fenix.FenixApplication
	at org.mozilla.fenix.ext.ContextKt.getApplication(Context.kt:29)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at io.mockk.proxy.jvm.advice.MethodCall.call(MethodCall.kt:14)
	at io.mockk.proxy.jvm.advice.SelfCallEliminatorCallable.call(SelfCallEliminatorCallable.kt:14)
	at io.mockk.impl.instantiation.JvmMockFactoryHelper.handleOriginalCall(JvmMockFactoryHelper.kt:83)
	at io.mockk.impl.instantiation.JvmMockFactoryHelper.access$handleOriginalCall(JvmMockFactoryHelper.kt:20)
	at io.mockk.impl.instantiation.JvmMockFactoryHelper$mockHandler$1$invocation$$inlined$stdFunctions$lambda$1.invoke(JvmMockFactoryHelper.kt:28)
	at io.mockk.impl.stub.MockKStub$handleInvocation$originalPlusToString$1.invoke(MockKStub.kt:227)
	at io.mockk.impl.stub.SpyKStub.defaultAnswer(SpyKStub.kt:15)
	at io.mockk.impl.stub.MockKStub.answer(MockKStub.kt:42)
	at io.mockk.impl.recording.states.AnsweringState.call(AnsweringState.kt:16)
	at io.mockk.impl.recording.CommonCallRecorder.call(CommonCallRecorder.kt:53)
	at io.mockk.impl.stub.MockKStub.handleInvocation(MockKStub.kt:263)
	at io.mockk.impl.instantiation.JvmMockFactoryHelper$mockHandler$1.invocation(JvmMockFactoryHelper.kt:25)
	at io.mockk.proxy.jvm.advice.Interceptor.call(Interceptor.kt:20)
	at org.mozilla.fenix.ext.ContextKt.getApplication(Context.kt:29)
	at org.mozilla.fenix.ext.ContextKt.getComponents(Context.kt:35)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at io.mockk.proxy.jvm.advice.MethodCall.call(MethodCall.kt:14)
	at io.mockk.proxy.jvm.advice.SelfCallEliminatorCallable.call(SelfCallEliminatorCallable.kt:14)
	at io.mockk.impl.instantiation.JvmMockFactoryHelper.handleOriginalCall(JvmMockFactoryHelper.kt:83)
	at io.mockk.impl.instantiation.JvmMockFactoryHelper.access$handleOriginalCall(JvmMockFactoryHelper.kt:20)
	at io.mockk.impl.instantiation.JvmMockFactoryHelper$mockHandler$1$invocation$$inlined$stdFunctions$lambda$1.invoke(JvmMockFactoryHelper.kt:28)
	at io.mockk.impl.stub.MockKStub$handleInvocation$originalPlusToString$1.invoke(MockKStub.kt:227)
	at io.mockk.impl.stub.SpyKStub.defaultAnswer(SpyKStub.kt:15)
	at io.mockk.impl.stub.MockKStub.answer(MockKStub.kt:42)
	at io.mockk.impl.recording.states.AnsweringState.call(AnsweringState.kt:16)
	at io.mockk.impl.recording.CommonCallRecorder.call(CommonCallRecorder.kt:53)
	at io.mockk.impl.stub.MockKStub.handleInvocation(MockKStub.kt:263)
	at io.mockk.impl.instantiation.JvmMockFactoryHelper$mockHandler$1.invocation(JvmMockFactoryHelper.kt:25)
	at io.mockk.proxy.jvm.advice.Interceptor.call(Interceptor.kt:20)
	at org.mozilla.fenix.ext.ContextKt.getComponents(Context.kt:35)
	at org.mozilla.fenix.components.toolbar.DefaultToolbarMenu.<init>(DefaultToolbarMenu.kt:405)
	at org.mozilla.fenix.toolbar.DefaultToolbarMenuTest.createMenu(DefaultToolbarMenuTest.kt:73)
	at org.mozilla.fenix.toolbar.DefaultToolbarMenuTest.WHEN the top toolbar is set THEN the first item in the list is the navigation(DefaultToolbarMenuTest.kt:107)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)

```

</details>

**Comment by jonalmeida** [comment]
> This pull request has conflicts when rebasing. Could you fix it @jonalmeida? 🙏

external-content duckduckgo com

**Review by rocketsroger — APPROVED** [review]
🚢

**Pull request merged**